File: SWEA/D4/1494. 사랑의 카운슬러/사랑의 카운슬러.cpp
#include <iostream>
#include <stdlib.h>

using namespace std;

pair<int, int> ary[20];
int n, fn;
long long res;

void sol(int idx, int state, int cnt) {
	if (idx + fn - cnt > n) return;
	if (idx == n) {
		long long f = 0, s = 0, temp;
		for (int i = 0; i < n; i++) {
			if (state & (1 << i)) {
				f += ary[i].first;
				s += ary[i].second;
			}
			else {
				f -= ary[i].first;
				s -= ary[i].second;
			}
		}
		temp = f * f + s * s;
		res = min(res, temp);
		return;
	}

	sol(idx + 1, state, cnt);
	if(cnt < fn)
		sol(idx + 1, state | (1 << idx), cnt + 1);
}

int main(void)
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int T, a, b;
	cin >> T;
	for (int tc = 1; tc <= T; tc++) {
		res = 800000000001;
		cin >> n;
		fn = n / 2;
		for (int i = 0; i < n; i++) {
			cin >> a >> b;
			ary[i] = { a, b };
		}
		sol(1, 1, 1);
		cout << "#" << tc << " " << res << "\n";
	}

	
	return 0;
}

File: SWEA/D4/1494. 사랑의 카운슬러/사랑의 카운슬러.java
import java.util.*;
import java.io.*;

public class Solution {
	static int ary[][];
	static int n, fn, cnt;
	static long res;
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		StringBuilder sb = new StringBuilder();
		int T = Integer.parseInt(br.readLine());
		
		ary = new int[20][2];
		for (int tc = 1; tc <= T; tc++) {
			res = Long.MAX_VALUE;
			n = Integer.parseInt(br.readLine());
			fn = n >> 1;
			for (int i = 0; i < n; i++) {
				st = new StringTokenizer(br.readLine());
				ary[i][0] = Integer.parseInt(st.nextToken());
				ary[i][1] = Integer.parseInt(st.nextToken());
			}
			cnt = 1;
			sol(1, ary[0][0], ary[0][1]);
			sb.append("#").append(tc).append(" ").append(res).append("\n");
		}
		System.out.print(sb);
	}
	
	static void sol(int idx, long f, long s) {
		if (idx + fn - cnt > n) return;
		if (idx == n) {
			long temp = f * f + s * s;
			res = Math.min(res, temp);
			return;
		}

		sol(idx + 1, f - ary[idx][0], s - ary[idx][1]);
		if (cnt < fn) {
			cnt++;
			sol(idx + 1, f + ary[idx][0], s + ary[idx][1]);
			cnt--;
		}
	}
}